src/util/generate_x509_certificate.c

```diff
@@ -0,0 +1,60 @@
+/******************************************************************************
+ *
+ * Copyright 2018 Xaptum, Inc.
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License
+ *
+ *****************************************************************************/
+#include <xtt.h>
+
+int generate_x509_certificate(const char *privkey_filename, const char *pubkey_filename, const char *id_filename, const char *certificate_filename)
+{
+    xtt_ecdsap256_pub_key pub;
+    xtt_ecdsap256_priv_key priv;
+    xtt_identity_type id = xtt_null_identity;
+
+    if (pubkey_filename != NULL && privkey_filename != NULL)
+    {
+        if(read_from_file(privkey_filename, priv.data, sizeof(xtt_ecdsap256_priv_key)) < 0){
+            return READ_FROM_FILE_ERROR;
+        }
+        if(read_from_file(pubkey_filename, pub.data, sizeof(xtt_ecdsap256_pub_key)) < 0){
+            return READ_FROM_FILE_ERROR;
+        }
+    } else
+    {
+        if (0 != xtt_crypto_create_ecdsap256_key_pair(&pub, &priv)) {
+            return -1;
+        }
+    }
+
+    unsigned char cert_buf[XTT_X509_CERTIFICATE_LENGTH];
+    if(id_filename == NULL)
+    {
+        id = xtt_null_identity;
+    } else
+    {
+        if(read_from_file(id_filename, id.data, sizeof(xtt_identity_type)) < 0)
+            return READ_FROM_FILE_ERROR;
+    }
+
+    if (0 != xtt_x509_from_ecdsap256_keypair(&pub, &priv, &id, cert_buf, sizeof(cert_buf))) {
+        return CERT_CREATION_ERROR;
+    }
+
+    if (save_to_file(cert_buf, sizeof(cert_buf), certificate_filename) < 0) {
```

Comment:
Throughout these functions, when we check the return code from one of our functions, like here, we should save that return code and return it (in case, for example, `save_to_file` some day can return multiple potential return error codes). This way, we don't lose the information that the underlying function may be trying to tell us.

So, something like:
```
int save_ret =save_to_file(...
if (save_ret < 0) {
    return save_ret;
}
```